### PR TITLE
Added categories to media-bundle

### PR DESCRIPTION
**Pull request by dominikmatt**
| Q | A
| --- | ---
| Bug fix? | no
| New feature? | yes
| BC breaks? | no
| Deprecations? | no
| Fixed tickets | fixes #issuenum
| Related issues/PRs | #issuenum
| License | MIT
| Documentation PR | sulu/sulu-docs#prnum

#### What's in this PR?

- added categories to media edit
- filter medias with categories in smartcontent

#### To Do

- [ ] Create a documentation PR
- [x] update husky ( https://github.com/massiveart/husky/pull/644 )

### CODE REVIEW

**Comment by wachterjohannes** [comment]
@dominikmatt the changelog is missing

File: src/Sulu/Bundle/MediaBundle/Entity/FileVersion.php [hunk 1]
```diff
@@ -116,6 +116,12 @@ class FileVersion implements AuditableInterface
     private $properties = '{}';
 
     /**
+     * @var \Doctrine\Common\Collections\Collection
+     */
+    private $categories;
+
```
Comment (wachterjohannes) [inline src/Sulu/Bundle/MediaBundle/Entity/FileVersion.php]: remove this empty line

File: src/Sulu/Bundle/MediaBundle/Entity/FileVersion.php [hunk 2]
```diff
@@ -667,4 +674,35 @@ public function setProperties(array $properties)
 
         return $this;
     }
+
+    /**
+     * Add categories
+     *
+     * @param \Sulu\Bundle\CategoryBundle\Entity\Category $categories
```
Comment (wachterjohannes) [inline src/Sulu/Bundle/MediaBundle/Entity/FileVersion.php]: only Category with use

File: src/Sulu/Bundle/MediaBundle/Entity/FileVersion.php [hunk 3]
```diff
@@ -667,4 +674,35 @@ public function setProperties(array $properties)
...
+     * @return self
+     */
+    public function addCategory(\Sulu\Bundle\CategoryBundle\Entity\Category $categories)
```
Comment (wachterjohannes) [inline src/Sulu/Bundle/MediaBundle/Entity/FileVersion.php]: same here

File: src/Sulu/Bundle/MediaBundle/Entity/FileVersion.php [hunk 4]
```diff
@@ -667,4 +674,35 @@ public function setProperties(array $properties)
...
+     */
+    public function removeCategories()
+    {
+        $this->categories->clear();
+    }
+
+    /**
+     * Get categories
+     *
+     * @return \Doctrine\Common\Collections\Collection
```
Comment (wachterjohannes) [inline src/Sulu/Bundle/MediaBundle/Entity/FileVersion.php]: only Collection

File: src/Sulu/Bundle/MediaBundle/Resources/translations/sulu/backend.fr.xlf [hunk 5]
```diff
@@ -278,6 +278,10 @@
                 <source>sulu.media.formats</source>
                 <target>Formats des médias</target>
             </trans-unit>
+            <trans-unit id="243d88cf2f64352w337js55369f221238" resname="sulu.media.categories">
+                <source>sulu.media.categories</source>
+                <target>Categories</target>
```
Comment (wachterjohannes) [inline src/Sulu/Bundle/MediaBundle/Resources/translations/sulu/backend.fr.xlf]: are you sure this this is french?

File: src/Sulu/Bundle/MediaBundle/Entity/FileVersion.php [hunk 6]
```diff
@@ -667,4 +673,35 @@ public function setProperties(array $properties)
...
+     */
+    public function removeCategories()
+    {
+        $this->categories->clear();
+    }
+
+    /**
+     * Get categories.
+     *
+     * @return \Doctrine\Common\Collections\Collection
```
Comment (wachterjohannes) [inline src/Sulu/Bundle/MediaBundle/Entity/FileVersion.php]: only Collection here

File: src/Sulu/Bundle/MediaBundle/Entity/MediaDataProviderRepository.php [hunk 7]
```diff
@@ -111,6 +111,7 @@ public function appendJoins(QueryBuilder $queryBuilder, $alias, $locale)
             ->leftJoin($alias . '.files', 'file')
             ->leftJoin('file.fileVersions', 'fileVersion', 'WITH', 'fileVersion.version = file.version')
             ->leftJoin('fileVersion.tags', 'tag')
+            ->leftJoin('fileVersion.categories', 'categories')
```
Comment (wachterjohannes) [inline src/Sulu/Bundle/MediaBundle/Entity/MediaDataProviderRepository.php]: don`t we need the translations here?

File: src/Sulu/Bundle/MediaBundle/Media/Manager/MediaManager.php [hunk 8]
```diff
@@ -785,6 +787,26 @@ protected function setDataToMedia(Media $media, $data, $user)
                             $media->setType($type);
                         }
                         break;
+                    case 'categories':
+                        $categoryIds = [];
+                        $media->removeCategories();
+
+                        if (count($value)) {
```
Comment (wachterjohannes) [inline src/Sulu/Bundle/MediaBundle/Media/Manager/MediaManager.php]: why not combining this two foreach?

File: src/Sulu/Bundle/MediaBundle/Entity/FileVersion.php (hunk 1 above)
Comment (dominikmatt) [inline src/Sulu/Bundle/MediaBundle/Entity/FileVersion.php]: done

File: src/Sulu/Bundle/MediaBundle/Media/Manager/MediaManager.php (hunk 8 above)
Comment (dominikmatt) [inline src/Sulu/Bundle/MediaBundle/Media/Manager/MediaManager.php]: so we only need one select, if we combine it we need for every category one select

File: src/Sulu/Bundle/MediaBundle/Entity/MediaDataProviderRepository.php (hunk 7 above)
Comment (dominikmatt) [inline src/Sulu/Bundle/MediaBundle/Entity/MediaDataProviderRepository.php]: done

**Pull request merged**